### PR TITLE
Application api

### PR DESCRIPTION
**Pull request by vinayak-trivedi**
Date: 18/11/2023

----

## Issue Ticket Number:- 
- add issue ticket number here e.g #1282 

## Description: 
This PR creates three endpoints:
- GET: /applications: returns all the applications which are not accepted or rejected, and if userId is passed in the query param it return a application of a particular user
- POST: /applications: adds the application of a user to join RDS
- PATCH: /applications/:applicationId: updates a particular application, this will only be accessible to super_user,

Is Under Feature Flag 
- [] Yes
- [x] No

Database changes
- [x] Yes
- [ ] No

Breaking changes (If your feature is breaking/missing something please mention pending tickets)
- [ ] Yes
- [x] No

Is Development Tested?

- [x] Yes
- [ ] No

Tested in staging?

- [ ] Yes
- [x] No

**Api contract:** https://github.com/Real-Dev-Squad/website-api-contracts/pull/142
**Data model doc:** https://github.com/Real-Dev-Squad/website-data-models/pull/59/files

GET API response for all applications:
image

GET API response for a particular userId
image

GET API response for getting applications of a particular status
image

POST API for adding application
image

PATCH API response for updating an application
image

Test coverage

Controller file:
image

AuthorizeOwn middleware
image

Model file
image

Route file
image

### CODE REVIEW

**Comment by vinit717** [comment]
Please add local working proof via Postman for every route

**Comment by vinit717** [comment]
Please add test stats

File: test/fixtures/applications/applications.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      introduction:
+        "mattis aliquam faucibus purus in massa tempor nec feugiat nisl pretium fusce id velit ut tortor pretium viverra suspendisse potenti nullam ac tortor vitae purus faucibus ornare suspendisse sed nisi lacus sed viverra tellus in hac habitasse platea dictumst vestibulum rhoncus est pellentesque elit ullamcorper dignissim cras tincidunt lobortis feugiat vivamus at augue eget arcu dictum varius duis at consectetur lorem donec massa sapien faucibus et molestie ac feugiat sed lectus vestibulum mattis ullamcorper velit sed ullamcorper morbi tincidunt ornare massa eget egestas purus viverra accumsan in nisl nisi scelerisque eu ultrices vitae auctor eu augue ut lectus arcu bibendum at",
+      country: "India",
+      numberOfHours: 20,
+    },
+  ];
```
Comment (heyrandhir) [inline test/fixtures/applications/applications.ts:64]: Nit : can generate some meaningful text from chatgpt instead of some random text 😇

**Review by heyrandhir — CHANGES_REQUESTED** [review]
Kindly include unit tests and a route for `/applications/:id`. If time allows, consider implementing pagination for the GET all applications response.

File: controllers/applications.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    if (status) {
+      const applicationsWithStatus = await ApplicationModel.getApplicationsBasedOnStatus(status)
+      return res.json({
+        message: 'applications returned successfully!',
+        applications: applicationsWithStatus
+      })
+    }
+
+    const applications = await ApplicationModel.getAllApplications();
```
[the remark this answers is not in the document]
Comment (vinayak-trivedi) [reply controllers/applications.ts:31]: https://github.com/Real-Dev-Squad/website-backend/issues/1761 created a different ticket for it

**Review by heyrandhir — COMMENTED** [review]
Don't see unit tests for the model function @vinayak-trivedi

**Comment by vinayak-trivedi** [comment]
> Don't see unit tests for the model function @vinayak-trivedi

@heyrandhir as discussed, I will be adding pagination in separate PR, so there will be changes in models as well, so that is why not adding model test here

**Pull request closed without merging**